src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.cpp

```diff
@@ -16,33 +23,112 @@
+    const std::array<double, Dim>& vortex_mean_velocity,
+    const double vortex_strength) noexcept
+    : adiabatic_index_(adiabatic_index),
+      perturbation_amplitude_(perturbation_amplitude),
+      vortex_center_(vortex_center),
+      vortex_mean_velocity_(vortex_mean_velocity),
+      vortex_strength_(vortex_strength) {}
 
-void VortexPerturbation::apply(
-    const gsl::not_null<Scalar<DataVector>*> mass_density_source,
-    const gsl::not_null<tnsr::I<DataVector, 3>*> momentum_density_source,
-    const gsl::not_null<Scalar<DataVector>*> energy_density_source,
-    const Scalar<DataVector>& vortex_mass_density,
-    const tnsr::I<DataVector, 3>& vortex_momentum_density,
-    const Scalar<DataVector>& vortex_energy_density,
-    const Scalar<DataVector>& vortex_pressure,
-    const Scalar<DataVector>& vortex_velocity_z,
-    const Scalar<DataVector>& dz_vortex_velocity_z) noexcept {
-  get(*mass_density_source) =
-      get(vortex_mass_density) * get(dz_vortex_velocity_z);
+template <size_t Dim>
+void VortexPerturbation<Dim>::pup(PUP::er& p) noexcept {
+  p | adiabatic_index_;
+  p | perturbation_amplitude_;
+  p | vortex_center_;
+  p | vortex_mean_velocity_;
+  p | vortex_strength_;
+}
+
+template <>
+void VortexPerturbation<2>::apply() const noexcept {}
+
+template <>
+void VortexPerturbation<3>::apply(
+    const gsl::not_null<Scalar<DataVector>*> source_mass_density_cons,
+    const gsl::not_null<tnsr::I<DataVector, 3>*> source_momentum_density,
+    const gsl::not_null<Scalar<DataVector>*> source_energy_density,
+    const tnsr::I<DataVector, 3>& x, const double time) const noexcept {
+  Solutions::IsentropicVortex<3> vortex(
+      adiabatic_index_, vortex_center_, vortex_mean_velocity_,
+      perturbation_amplitude_, vortex_strength_);
+  const size_t size = get<0>(x).size();
+  const auto vortex_primitives = vortex.variables(
+      x, time,
+      tmpl::list<Tags::MassDensity<DataVector>, Tags::Velocity<DataVector, 3>,
+                 Tags::SpecificInternalEnergy<DataVector>,
+                 Tags::Pressure<DataVector>>{});
+
+  Variables<
+      tmpl::list<::Tags::TempScalar<0>, ::Tags::TempI<1, 3, Frame::Inertial>,
+                 ::Tags::TempScalar<2>>>
+      temp_buffer(size);
+  auto& vortex_mass_density_cons = get<::Tags::TempScalar<0>>(temp_buffer);
+  auto& vortex_momentum_density =
+      get<::Tags::TempI<1, 3, Frame::Inertial>>(temp_buffer);
+  auto& vortex_energy_density = get<::Tags::TempScalar<2>>(temp_buffer);
+
+  NewtonianEuler::ConservativeFromPrimitive<3>::apply(
+      make_not_null(&vortex_mass_density_cons),
+      make_not_null(&vortex_momentum_density),
+      make_not_null(&vortex_energy_density),
+      get<Tags::MassDensity<DataVector>>(vortex_primitives),
+      get<Tags::Velocity<DataVector, 3, Frame::Inertial>>(vortex_primitives),
+      get<Tags::SpecificInternalEnergy<DataVector>>(vortex_primitives));
+
+  DataVector dz_vortex_velocity_z =
```

Comment:
save an allocation by storing `dz_vortex_velocity_z` in `*source_mass_density_cons`, and then do the multiplication by `vortex_mass_density_cons` after computing the other sources